### ISSUE TITLE
[Discussion] Improve the Code Editor

### ISSUE THREAD

**Issue by carlsmith** [nesbox/TIC-80#723]
I quite like the current code editor, but it's still needs fixing. I'll explain why in the comments, so you can skip that bit if you already agree.

Thinking about TIC-80 like an OS, we could treat the bar along the top of the UI as separate from the rest of the UI, a bit like the menubar in OS X.

OS X uses the menubar to merge the system menu with the currently focussed window's menu to save space, which is similar to what TIC-80 does.

## Text and Graphics Modes

If TIC-80 had a *graphics mode* (which was just how it works now) and a *text mode* (with a higher resolution, but limited to a grid of glyphs, each with a foreground and background colour), then the console and editor could use the text mode, and everything else continue to use the graphics mode.

TIC-80 could define a *desktop mode* (which just means that the menubar is visible at the top) and *fullscreen mode* (without the menubar).

Assuming we want to keep everything else the same:

- console: text mode in fullscreen
- code editor: text mode on desktop
- other editors: graphics mode on desktop
- running games and demos: graphics mode in fullscreen.

## The Numbers

To get the text mode, the 240x136 resolution could be doubled to 480x272, which would allow 80 characters if they were 6px wide (really 5px, with a 1px space). The text mode resolution could be tripled to 720x408 (offering 9px per character, which seems ideal).

Having the text mode use a multiple of the graphics mode will everything still look pixel perfect when the (graphics mode) menubar is above the (text mode) code editor.

A TIC-80 text mode should probably use memory mapped glyphs too.

**Comment by carlsmith** [comment]
The reasons why I think external editors are a problem...

If a user asked for GIMP integration in TIC-80, because they found the built in sprite editor too limiting, they'd be told to use a different tool or a workaround. The same goes for tile maps and audio, and should apply to text editors equally. You can argue that programmers are particular about their tools, but pixel artists and chiptune musicians are just the same.

The problem (in my opinion) is just that PICO-8 designed an editor that is *too limiting*, and then allowed external editors, because they couldn't think of a solution. Other fantasy consoles copied the same approach, and ended up with the same compromise.

External editors break the sense of actually working on a fantasy console (ruining the suspension of disbelief). They're clunky (you can't hit Escape to get to the shell, or Ctrl-Enter to run the code, F2, F3 etc). They force new programmers to look at one set of docs for the TIC-80 computer, and another set for their editor. (complicating things for educators and tech writers too), and a lot of new programmers use TIC-80 in their browser, where they don't have another editor.

**Comment by nesbox** [comment]
I think we could double the resolution to 480x272 decreasing the color bits from 4bpp to 2bpp in the Code Editor. This won't break the rules and still fit VRAM size. So, we can move in this direction...
Interesting ideas, thanks.

**Comment by carlsmith** [comment]
Awesome. Thank you. You're the best.

**Comment by ephetic** [comment]
How about a few enhancements like highlighting matching bracket and rudimentary completion? I'll research the history, but I wouldn't be surprised if they existed on Vic, Amiga, etc.

**Comment by Anrock** [comment]
@ephetic matching brackets were merged couple of days ago: https://github.com/nesbox/TIC-80/issues/1075

**Comment by ephetic** [comment]
Whoo hoo!

**Comment by mttdr** [comment]
Hello, was this idea abandoned?
Having a comfier code editor would make tic80 even better, and I do agree that switching over external text editors makes for a worse workflow.

**Comment by joshgoebel** [comment]
> Hello, was this idea abandoned?

Which idea exactly?  We're certainly probably not going to double the resolution anytime soon.

> I do agree that switching over external text editors makes for a worse workflow.

It would be nice to understand HOW it makes for a worse workflow... as the internal editor will NEVER be as good as the best external editors.

**Comment by mttdr** [comment]
Yes I was referring to the idea of doubling the code editor resolution.
Of course it will never be as good as full fledged external IDEs, but it already has a lot of useful features (goto line, functions outline, and find really go a long way to move around the code)
Increasing the resolution seems very promising because it would allow for:
- longer lines to be readable on screen
- more readable font (as already linked above https://github.com/nesbox/TIC-80/issues/723#issuecomment-896782300)
- more room in the UI to implement other features down the line (eg tabs)

> It would be nice to understand HOW it makes for a worse workflow

A big part of the appeal of the fantasy computer is to use a single and defined set of tools, that although limited has anything you need to be creative. In reality, the code editor can feel "claustrophobic" as soon as you write more than a couple of functions and start using descriptive (and long) variable names, because you end up always scrolling around to read the full context.

As described at the beginning of the thread (https://github.com/nesbox/TIC-80/issues/723#issuecomment-419683436), switching over to external editors breaks the suspension of disbelief of using a fantasy computer.
More practically it puts new users or students to the choice of having to choose a new tool.
To experienced users i think it's still annoying to switch back and forth, as you always need to save every change, assure the external editor reloads the changes done in TIC-80, plus the external editor might miss the auto-completion of TIC-80 specific functions, and so on...

**Comment by joshgoebel** [comment]
Ok, I understand that POV - that staying in the "fantasy computer" world could be an enjoyable thing... sadly I just can't imagine nesbox changing the resolution... it's too baked into what the platform is - it's hardware with a low-resolution screen... that's part of the fantasy. :-)

**Comment by mttdr** [comment]
I see, I was curious because @nesbox seemed to like the idea (back in 2018 at least), so I wondered if this was still in the todo backlog or just abandoned for whatever reason.

**Comment by joshgoebel** [comment]
Oh? Do you have a link?  Would love to see that.  He's pretty open to most ideas, but he's been pretty closed (in my experience) on the core capabilities of the hardware.

**Comment by mttdr** [comment]
https://github.com/nesbox/TIC-80/issues/723#issuecomment-419699952
here, it's up in this same thread

**Comment by jpfed** [comment]
The idea that fantasy consoles should remain low-resolution is understandable. I just wanted to note that the computers I happened to grow up with (Apple II, CGA- and VGA- era PCs) also had the ability to switch among resolutions, including higher-resolution/lower color modes. My particular nostalgia / suspension of disbelief does not depend on remaining at a single resolution/color mode- and I do wish for a less claustrophobic editing experience.

**Comment by joshgoebel** [comment]
> here, it's up in this same thread

Indeed.  Wow... though he did say it would have to be 4 colors (2bit)... so given one background color now that's only 3 colors... one for text... two for syntax highlighting... seems a bit limited to me...  adding a text/console mode that was character based would be another idea here... (if we're just willing to accept anything try to original hardware of a certain era)... the ram would be just rows * cols (ASCII) + rows * cols (attributes)... 80x25 would be only 4kb (16 bit color, no effects)... or 6kb if we allowed 16 bit attributes (bold, italic, underline, invert, etc)... so we actually have tons of RAM in that case...
